### PR TITLE
Optimization of `__serial_merge` function

### PR DESCRIPTION
**Pull request by SergeyKopienko**
In this PR we make optimization of `__serial_merge` function.
The justification: previous implementation is too complicated:
- a lot of conditions;
- a lot of for-loops (and one inside another).

With this new implementation of `__serial_merge` function we have significant performance profit for `merge` and `merge_sort` algorithms on all data sizes.

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h [hunk 1]
```diff
@@ -80,52 +80,35 @@ __find_start_point(const _Rng1& __rng1, const _Rng2& __rng2, const _Index __i_el
...
+
+    const auto __rng1_idx_end = __start1 + __rng1_size;
+    const auto __rng2_idx_end = __start2 + __rng2_size;
+    const auto __rng3_idx_end = __start3 + __rng3_size;
+
+    _Index __rng1_idx = __start1;
+    _Index __rng2_idx = __start2;
+    _Index __rng3_idx = __start3;
+    
+    bool __rng1_idx_less__n1, __rng2_idx_less__n2;
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:100]: The uninitialized variables...  I guess we can move the variables definition  to the place where ones are initialized within the loop, w/o performance any penalty.

**Review by MikeDvorskiy — COMMENTED** [review]
Looks very good to me, excepting one minor comment, see above/

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h [hunk 2]
```diff
@@ -272,17 +274,17 @@ struct __merge_sort_global_submitter<_IndexT, __internal::__optional_kernel_name
                             const oneapi::dpl::__ranges::drop_view_simple __rng2(__dst, __offset + __n1);
 
                             const auto start = __find_start_point(__rng1, __rng2, __i_elem_local, __n1, __n2, __comp);
-                            __serial_merge(__rng1, __rng2, __rng /*__rng3*/, start.first, start.second, __i_elem,
-                                           __chunk, __n1, __n2, __comp);
+                            __serial_merge<__gpu_chunk>(__rng1, __rng2, __rng /*__rng3*/, start.first, start.second,
```
Comment (mmichel11) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h:277]: I think renaming `__gpu_chunk` to something like `__unroll_factor` is more clear since there is a different `__chunk` and the device may not be a GPU.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 1 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:100]: fixed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h (hunk 2 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h:277]: Semantically in this code the variable exactly mean the value of `chunk` for any device that can also be used to accelerate a 3D graphics API. In other words - the chunk for non-CPU device.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h [hunk 3]
```diff
@@ -96,8 +97,8 @@ __serial_merge(const _Rng1& __rng1, const _Rng2& __rng2, _Rng3& __rng3, _Index _
     _Index __rng1_idx = __start1;
     _Index __rng2_idx = __start2;
     _Index __rng3_idx = __start3;
-    
-    #pragma unroll _UnrollFactor
+
+#pragma unroll _UnrollFactor
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:101]: `#pragma` is not portable.. OneDPL has a macro
// Portability "#pragma" definition
```
#ifdef _MSC_VER
#    define _ONEDPL_PRAGMA(x) __pragma(x)
#else
#    define _ONEDPL_PRAGMA(x) _Pragma(#    x)
#endif

```

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:101]: Could you please explain how to apply this for unroll?

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: We made decision together with @dmitriy-sobolev to remove `unroll` here at all: no perf profit has been detected.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h [hunk 4]
```diff
@@ -85,47 +85,27 @@ void
...
+    const _Index __rng2_size = std::min<_Index>(__n2 > __start2 ? __n2 - __start2 : (_Index)0, (_Index)__chunk);
+    const _Index __rng3_size = std::min<_Index>(__rng1_size + __rng2_size, (_Index)__chunk);
+
+    const auto __rng1_idx_end = __start1 + __rng1_size;
+    const auto __rng2_idx_end = __start2 + __rng2_size;
+    const auto __rng3_idx_end = __start3 + __rng3_size;
+
+    _Index __rng1_idx = __start1;
+    _Index __rng2_idx = __start2;
+    _Index __rng3_idx = __start3;
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:98]: `__rng3_idx` is redefined in the loop below.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h [hunk 5]
```diff
@@ -85,47 +85,27 @@ void
...
-            __rng3[__start3 + __i] = __rng1[__start1 + __i];
-    }
-    else
+    const _Index __rng1_size = std::min<_Index>(__n1 > __start1 ? __n1 - __start1 : (_Index)0, (_Index)__chunk);
+    const _Index __rng2_size = std::min<_Index>(__n2 > __start2 ? __n2 - __start2 : (_Index)0, (_Index)__chunk);
+    const _Index __rng3_size = std::min<_Index>(__rng1_size + __rng2_size, (_Index)__chunk);
+
+    const auto __rng1_idx_end = __start1 + __rng1_size;
+    const auto __rng2_idx_end = __start2 + __rng2_size;
+    const auto __rng3_idx_end = __start3 + __rng3_size;
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:94]: Let's use `_Index` instead of auto, because the type is known.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h [hunk 6]
```diff
@@ -85,47 +85,27 @@ void
...
-                __rng3[__start3 + __i] = __val1;
-                if (++__start1 == __n1)
-                {
-                    //copying a residual of the second seq
-                    for (++__i; __i < __chunk && __start2 < __n2; ++__i, ++__start2)
-                        __rng3[__start3 + __i] = __rng2[__start2];
-                }
-            }
-        }
+        const bool __rng1_idx_less__n1 = __rng1_idx < __rng1_idx_end;
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:102]: A nitpick: `__rng1_idx_less__n1` -> `__rng1_idx_less_n1`. Another option is `__rng1_idx_valid`.

**Review by MikeDvorskiy — APPROVED** [review]
Looks very good to me.

**Review by dmitriy-sobolev — COMMENTED** [review]
It also looks good to me. Below are some minor suggestions.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h [hunk 7]
```diff
@@ -85,47 +85,27 @@ void
...
+        const bool __rng2_idx_less__n2 = __rng2_idx < __rng2_idx_end;
```
Comment (akukanov) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: Can it happen that both these conditions are false at the same time? The code that follows does not seem to handle such case correctly.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 6 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: Fixed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 5 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: Fixed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 4 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: Fixed

**Review by dmitriy-sobolev — APPROVED** [review]

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h [hunk 8]
```diff
@@ -85,47 +85,26 @@ void
...
-    {
-        //copying a residual of the first seq
-        const _Index __n = std::min<_Index>(__n1 - __start1, __chunk);
-        for (std::uint8_t __i = 0; __i < __n; ++__i)
-            __rng3[__start3 + __i] = __rng1[__start1 + __i];
-    }
-    else
+    const _Index __rng1_size = std::min<_Index>(__n1 > __start1 ? __n1 - __start1 : (_Index)0, (_Index)__chunk);
+    const _Index __rng2_size = std::min<_Index>(__n2 > __start2 ? __n2 - __start2 : (_Index)0, (_Index)__chunk);
+    const _Index __rng3_size = std::min<_Index>(__rng1_size + __rng2_size, (_Index)__chunk);
```
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:90]: Is there a reason to do C style casting here rather than doing explicit construction of a `_Index` type rvalue (or perhaps a `static_cast`)?

`(_Index)0` -> `_Index{0}`

I don't think we do a lot of c style casting in the repository.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h [hunk 9]
```diff
@@ -85,47 +85,26 @@ void
...
-                        __rng3[__start3 + __i] = __rng2[__start2];
-                }
-            }
-        }
+        const bool __rng1_idx_less_n1 = __rng1_idx < __rng1_idx_end;
+        const bool __rng2_idx_less_n2 = __rng2_idx < __rng2_idx_end;
+
+        __rng3[__rng3_idx] =
+            __rng1_idx_less_n1 && __rng2_idx_less_n2
+                ? (__comp(__rng2[__rng2_idx], __rng1[__rng1_idx]) ? __rng2[__rng2_idx++] : __rng1[__rng1_idx++])
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:106]: Postfix increment leads to a temporary variable creation usually...

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h [hunk 10]
```diff
@@ -85,47 +85,27 @@ void
...
-    {
-        //copying a residual of the first seq
-        const _Index __n = std::min<_Index>(__n1 - __start1, __chunk);
-        for (std::uint8_t __i = 0; __i < __n; ++__i)
-            __rng3[__start3 + __i] = __rng1[__start1 + __i];
-    }
-    else
+    const _Index __rng1_size = std::min<_Index>(__n1 > __start1 ? __n1 - __start1 : (_Index)0, (_Index)__chunk);
+    const _Index __rng2_size = std::min<_Index>(__n2 > __start2 ? __n2 - __start2 : (_Index)0, (_Index)__chunk);
+    const _Index __rng3_size = std::min<_Index>(__rng1_size + __rng2_size, (_Index)__chunk);
```
Comment (akukanov) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: Since `uint8_t` is no more used for iteration, maybe is it better to change the `__chunk` parameter type to `_Index` instead of explicit conversions?
Also, `_Index(0)` is slightly better than the C-style cast.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 9 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:106]: Post-increment here required from logic side of implementation: we should get value from range and then increment index.  And exactly increment index for that range from which we got value.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 8 above)
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:90]: I agree they should result in the same code, I point it out because it is against what I understand to be our convention, but its a nitpick and I wont insist.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 5 above)
Comment (akukanov) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: Both `__startX` and `__rngX_size` are of type `_Index`, so what is `auto` for?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 8 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: `_Index{0}` looks good.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 9 above)
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:106]: > Post-increment here required from logic side of implementation: we should get value from range and then increment index. And exactly increment index for that range from which we got value.

Of course I understand the logic here and I am aware of difference between Post-increment  and Pre-increment  operations in C/C++...

I am talking about possible additional temporary variables creation. Usually it is bad style in C/C++ by performance reasons. So, do we have / might have any performance penalty comparing with code w/o temporary variable in given context?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 7 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: I believe it's not possible due 
1) condition in the for-loop: `__rng3_idx < __rng3_idx_end`
2) and how we define the value of `__rng3_idx_end` :
```C++
    const _Index __rng1_size = std::min<_Index>(__n1 > __start1 ? __n1 - __start1 : _Index{0}, (_Index)__chunk);
    const _Index __rng2_size = std::min<_Index>(__n2 > __start2 ? __n2 - __start2 : _Index{0}, (_Index)__chunk);
    const _Index __rng3_size = std::min<_Index>(__rng1_size + __rng2_size, (_Index)__chunk);
```

So, logically we should fill all required data and this mean we have them in the first of in the second ranges.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 9 above)
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:106]: @MikeDvorskiy  I suppose the suggestion would then be to convert the ternary operators into a couple `if`/`else` statements where we can group the increment in a separate expression in the if block rather than embedding it in the assignment.
Is that your suggestion?

Comment (akukanov) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:106]: Since `operator[]` here is likely a function call, and the expressions for the arguments of function calls are evaluated before the call, a temporary variable is technically needed. On the other hand, arguments are passed to the function calls either on the stack or in registers (depending on the calling convention), so the place for that temporary variable is likely to exist anyway.

Overall, I would not bother with such micro-optimizations, given that even explicit loop unrolling turned out to have no notable impact.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h [hunk 11]
```diff
@@ -85,47 +85,27 @@ void
...
+
+    const auto __rng1_idx_end = __start1 + __rng1_size;
+    const auto __rng2_idx_end = __start2 + __rng2_size;
+    const auto __rng3_idx_end = __start3 + __rng3_size;
+
+    _Index __rng1_idx = __start1;
+    _Index __rng2_idx = __start2;
+    _Index __rng3_idx = __start3;
+
+    for (_Index __rng3_idx = __start3; __rng3_idx < __rng3_idx_end; ++__rng3_idx)
```
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: My personal preferences - to define in the for-loop only one variable to make for-loop more clear.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 7 above)
Comment (akukanov) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: Perhaps add a comment that at least one of the conditions is (expectedly) true.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 9 above)
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:106]: > Is that your suggestion?

Not necessary  `if/else`.... for example: `(++__rng2_idx, __rng2[__rng2_idx-1])`

```
__rng3[__rng3_idx] =
            __rng1_idx_less__n1 && __rng2_idx_less__n2
                ? (__comp(__rng2[__rng2_idx], __rng1[__rng1_idx]) ? (++__rng2_idx, __rng2[__rng2_idx-1]) : (++__rng1_idx, __rng1[__rng1_idx-1]))
                : (__rng1_idx_less__n1 ? (++__rng1_idx, __rng1[__rng1_idx-1]) : (++__rng2_idx, __rng2[__rng2_idx-1]));
```

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 10 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: Fixed.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (akukanov) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: OK, I have no desire to confront your personal preferences :)

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 9 above)
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:109]: > Overall, I would not bother with such micro-optimizations, given that even explicit loop unrolling turned out to have no notable impact.

I just ask a question about post-increment, nothing more :)
Additionally, I've just compared the asm files with -03. The are identical. 

So, Q.E.D. `we do believe in a compiler.`, This is something like what I expected to hear as an answer, not explanation how serial merge works :)

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 7 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: Done.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 9 above)
Comment (akukanov) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:109]: > for example: (++__rng2_idx, __rng2[__rng2_idx-1])
... where `__rng2_idx-1` would be I believe an rvalue similar to the one created by `__rng2_idx++` :)

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h [hunk 12]
```diff
@@ -83,49 +83,28 @@ __find_start_point(const _Rng1& __rng1, const _Rng2& __rng2, const _Index __i_el
...
-                }
-            }
-        }
+        const bool __rng1_idx_less_n1 = __rng1_idx < __rng1_idx_end;
+        const bool __rng2_idx_less_n2 = __rng2_idx < __rng2_idx_end;
+
+        __rng3[__rng3_idx] =
+            __rng1_idx_less_n1 && __rng2_idx_less_n2
+                ? (__comp(__rng2[__rng2_idx], __rng1[__rng1_idx]) ? __rng2[__rng2_idx++] : __rng1[__rng1_idx++])
+                : (__rng1_idx_less_n1 ? __rng1[__rng1_idx++] : __rng2[__rng2_idx++]);
```
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:110]: So, I don't know if its worth the cost that may make it harder to read, but technically there are only 2 outcomes:  
use rng1 and increment its index 
use rng2 and increment its index

We should be able to form a single ternary / if statement to get the result we want.  Let me give it a try (obviously it'll need formatting):
```suggestion
        __rng3[__rng3_idx] =
            ((__rng1_idx_less_n1 && __rng2_idx_less_n2 && __comp(__rng2[__rng2_idx], __rng1[__rng1_idx])) || __rng2_idx_less_n2) ? __rng2[__rng2_idx++] : __rng1[__rng1_idx++]);
```

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:110]: Its possible the compiler can do this for us but if not... we have an extra unnecessary branch which is costly.

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:110]: I believe technically you suggestion is absolutely correct. but a little bit not so readable.
@akukanov, @MikeDvorskiy, @dmitriy-sobolev what do you think?

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:110]: I think at the benefit of removing a branch in an inner loop of the main computation loop of an algorithm, we can sacrifice readability.  If we feel its confusing, lets add a comment to clarify.

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:110]: One more variant:
```C++
        __rng3[__rng3_idx] =
            ((__rng1_idx_less_n1 && __comp(__rng2[__rng2_idx], __rng1[__rng1_idx])) && __rng2_idx_less_n2)
                ? __rng2[__rng2_idx++] : __rng1[__rng1_idx++]);
```

Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:110]: As for me, any variant readable...
But I would prefer variant w/o redundant braces, it makes the code lines less readable for me.
Regarding formatting - it doesn't matter for me with readable perspective.

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:110]: But `__rng1_idx_less_n1==false` and `__rng2_idx_less_n2== true`, this variant will attempt to use `__rng1` and crash, wont it?

Comment (akukanov) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:110]: I think Dan's code would not work as needed when `__rng1_idx_less_n1 == true && __rng2_idx_less_n2 == true && __comp(__rng2[__rng2_idx], __rng1[__rng1_idx]) == false`.

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:110]:
```
// If we are within both range 1 and 2, and the element from range2 is "less than" the element from range2, use it.  
// If we are not within both ranges, but we are still within range2, use the element from range2
// Otherwise, we know we are within range1 because we are in the loop, so use the element from range1.
```

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:110]: @akukanov Yeah, I think you are right. Sorry let me fix the check, it has to get a bit more complex.

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:110]: Second try:
```suggestion
        __rng3[__rng3_idx] =
            ((__rng1_idx_less_n1 && __rng2_idx_less_n2 && __comp(__rng2[__rng2_idx], __rng1[__rng1_idx])) || !__rng1_idx_less_n1) ? __rng2[__rng2_idx++] : __rng1[__rng1_idx++]);
```
This is what I originally had but thought I could get clever to avoid the negation.  I think this resolves the issues though.

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:110]: One comma pairs aren't required I believe:
```C++
        __rng3[__rng3_idx] =
            (__rng1_idx_less_n1 && __rng2_idx_less_n2 && __comp(__rng2[__rng2_idx], __rng1[__rng1_idx])) || !__rng1_idx_less_n1
                ? __rng2[__rng2_idx++] : __rng1[__rng1_idx++];
```

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: Yeah that is fine.  I personally like to wrap the condition in ternary operations in parens just to make the order of operations by the compiler clear both to the reader and the compiler, but I don't think it is necessary for correctness.

`!__rng1_idx_less_n1 ? __rng2[__rng2_idx++] : __rng1[__rng1_idx++]` alone is a "valid" expression, which could then be combined via `||` with the first half of the conditional, so for me it is less confusing to just explicitly group the condition in parens to remove any doubt.

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: Applied.

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: > Yeah that is fine. I personally like to wrap the condition in ternary operations in parens just to make the order of operations clear both to the reader and the compiler, but I don't think it is necessary for correctness.
> 
> `!__rng1_idx_less_n1 ? __rng2[__rng2_idx++] : __rng1[__rng1_idx++]` alone is a "valid" expression, which could then be combined via `||` with the first half of the conditional, so for me it is less confusing to just explicitly group the condition in parens to remove any doubt.

ok, applied too.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h]: May I resolve this conversation?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h [hunk 13]
```diff
@@ -63,7 +63,7 @@ struct __group_merge_path_sorter
     template <typename _StorageAcc, typename _Compare>
     bool
     sort(const sycl::nd_item<1>& __item, const _StorageAcc& __storage_acc, _Compare __comp, std::uint32_t __start,
-         std::uint32_t __end, std::uint32_t __sorted, std::uint16_t __data_per_workitem,
+         std::uint32_t __end, std::uint32_t __sorted, std::uint32_t __data_per_workitem,
```
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h:66]: Last minor question:  
What changed here to need this type change?  Is it a bug fix?  it seems unrelated to the rest of the PR.

**Review by mmichel11 — DISMISSED** [review]
LGTM on my side.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h [hunk 14]
```diff
@@ -82,50 +82,33 @@ __find_start_point(const _Rng1& __rng1, const _Rng2& __rng2, const _Index __i_el
...
-                    //copying a residual of the second seq
-                    for (++__i; __i < __chunk && __start2 < __n2; ++__i, ++__start2)
-                        __rng3[__start3 + __i] = __rng2[__start2];
-                }
-            }
-        }
+        const bool __rng1_idx_less_n1 = __rng1_idx < __rng1_idx_end;
+        const bool __rng2_idx_less_n2 = __rng2_idx < __rng2_idx_end;
+
+        // One of __rng1_idx_less_n1 and __rng2_idx_less_n2 should be true here
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:104]: Do we whether want to write an `assert` here?

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:104]: It's Kernel code, so no asserts.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h [hunk 15]
```diff
@@ -82,50 +82,33 @@ __find_start_point(const _Rng1& __rng1, const _Rng2& __rng2, const _Index __i_el
...
+        // because 1) we should fill output data with elements from one of the input ranges
+        // 2) we calculate __rng3_idx_end as std::min<_Index>(__rng1_size + __rng2_size, __chunk).
+        __rng3[__rng3_idx] =
+            ((__rng1_idx_less_n1 && __rng2_idx_less_n2 && __comp(__rng2[__rng2_idx], __rng1[__rng1_idx])) ||
+             !__rng1_idx_less_n1)
+                ? __rng2[__rng2_idx++]
+                : __rng1[__rng1_idx++];
```
Comment (MikeDvorskiy) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:111]: common! it is too much.. Let's not writting the code in one column :)

`!__rng1_idx_less_n1)  ? __rng2[__rng2_idx++]  : __rng1[__rng1_idx++];`

is absolutely readable.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 14 above)
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:104]: 1) CPU device?
2) https://intel.github.io/llvm/design/Assert.html
3) I guess we are going to share the function with the host code as well, probably...

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:104]: As noted @akukanov earlier in https://github.com/oneapi-src/oneDPL/pull/1933 conversation, we can't use `assert` calls inside Kernel code. So what you propose exactly?

**Review by danhoeflinger — APPROVED** [review]
LGTM pending green CI, thanks

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h (hunk 13 above)
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h:66]: After offline discussion, we see that `__data_per_workitem` is only used in 32 bit unsigned context, and we have made things more explicit by removing `auto` in the return from `__find_start_point` for the other parameters with `_Index` type going into `__serial_merge` so that it is clear that we will end up with a consistent type.  
Thanks for the changes here.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 15 above)
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:111]: "disagree and commit"

**Review by mmichel11 — DISMISSED** [review]

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 15 above)
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:111]: > It's automatic format.

What kind of "automatic" ? by IDE/text redactor or by clang format (form CI) ?

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:111]: It's someone IDE with our clang-format setting.
Is it really important?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 14 above)
Comment (akukanov) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:104]: For the sake of accuracy, I have said this:
> We should consider the use of assert in device code as an extension of DPC++ and possibly other SYCL implementations, so it should be guarded by some macros.

Practically, I would perhaps recommend to use the assertion during development to ensure that it does not fail for our tests, but then remove it. The rationale is that any later changes to the code would likely require the assertion to be modified as well, so it does not seem much useful beyond the initial testing.

**Review by dmitriy-sobolev — DISMISSED** [review]

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h (hunk 14 above)
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h:104]: Based on [1933](https://github.com/oneapi-src/oneDPL/pull/1933#discussion_r1873731880) discussion,  I agree with following conclusion: 

`We should consider the use of assert in device code as an extension of DPC++ and possibly other SYCL implementations, so it should be guarded by some macros`

I don't insist doing it in this PR but I will create an GitHub issue.

**Comment by SergeyKopienko** [comment]
Unfortunately clang-formatting has been fixed. So approves required again.
Sorry for inconveniences.

**Review by danhoeflinger — APPROVED** [review]
